File: src/python/bsp.py
import util
import draw
import wad

#def draw_bsp_nodes(nodes, surf):
#    root_node = get_root_node(nodes)

#    draw_bsp_node(root_node, surf)


def is_ssect_idx(idx):
    return (idx & (1<<15)) != 0


def is_on_left(node, x, y):
    dx = x - node.partition_x_coord
    dy = y - node.partition_y_coord

    return (((dx * node.dy) - (dy * node.dx)) <= 0)


def get_real_idx(idx):
    masked_idx = (idx & (~(1<<15)))
    return (idx & (~(1<<15))) 

def traverse_bsp_node(node_idx,
                      node_depth,
                      level_data,
                      node_callback,
                      ssector_callback):

    
    if is_ssect_idx(node_idx):
        ssector_callback(level_data['SSECTORS'][get_real_idx(node_idx)])
    else:
        node = level_data['NODES'][get_real_idx(node_idx)]
        node_callback(node, node_depth)
        traverse_bsp_node(node.left_child, node_depth+1,
                          level_data, node_callback, ssector_callback)
        traverse_bsp_node(node.right_child, node_depth+1,
                          level_data, node_callback, ssector_callback)

    
        
def traverse_all_bsp_nodes(level_data,
                           node_callback=lambda node: None,
                           ssect_callback=lambda ssect: None):
    nodes = level_data['NODES']
    root_idx = len(nodes)-1
    traverse_bsp_node(root_idx, 0, level_data, node_callback, ssect_callback)


def bsp_child_onscreen(node, left_child=True):
    
    if left_child:
        v1 = (node.left_box_left, node.left_box_top)
        v2 = (node.left_box_right, node.left_box_top)
        v3 = (node.left_box_right, node.left_box_bottom)
        v4 = (node.left_box_left, node.left_box_bottom)
    else:
        v1 = (node.right_box_left, node.right_box_top)
        v2 = (node.right_box_right, node.right_box_top)
        v3 = (node.right_box_right, node.right_box_bottom)
        v4 = (node.right_box_left, node.right_box_bottom)


    return draw.bbox_on_screen(v1,v2,v3,v4)


def get_bsp_tree_bounds(level_data):
    nodes = level_data['NODES']
    root_node_idx = len(nodes)-1
    node = nodes[root_node_idx]

    xs = [
        node.left_box_left, node.left_box_right,
        node.right_box_left, node.right_box_right,
    ]

    ys = [
        node.left_box_top, node.left_box_bottom,
        node.right_box_top, node.right_box_bottom,
    ]
    
    return [min(xs), max(xs), min(ys), max(ys)]
    
    
def traverse_bsp_front_to_back(level_data,
                               x, y, start_subsector,
                               node_callback=lambda node,visiting_left, visiting_right: None,
                               ssect_callback=lambda ssect: None):
    nodes = level_data['NODES']
    ssectors = level_data['SSECTORS']
    root_node_idx = len(nodes)-1
    #depth = 0

    #stk = [root_node_idx]


    src_sector_idx = wad.get_subsector_sector_idx(start_subsector, level_data)
    
    def recurse(node_idx):
        if is_ssect_idx(node_idx):
            subsector = ssectors[get_real_idx(node_idx)]
            dest_sector_idx = wad.get_subsector_sector_idx(subsector, level_data)
            if True: #((not draw.pvs_check) or
                # (draw.pvs_check and wad.maybe_line_of_sight(src_sector_idx, dest_sector_idx, level_data))):
                real_idx = get_real_idx(node_idx)
                ssect_callback(real_idx, ssectors[real_idx])
            return

        node = nodes[get_real_idx(node_idx)]

        on_left = is_on_left(node, x, y)
        visiting_left = False
        visiting_right = False

        second_child = None
        frustum_culling = draw.bsp_node_frustum_culling
        
        if on_left:
            first_child = node.left_child
            visiting_left = True
            if (not frustum_culling) or bsp_child_onscreen(node, left_child=False):
                visiting_right = True
                second_child = node.right_child
                                

        else:
            first_child = node.right_child
            visiting_right = True
            if (not frustum_culling) or bsp_child_onscreen(node, left_child=True):
                visiting_left = True
                second_child = node.left_child

                

        
        if node_callback:
            node_callback(node, visiting_left, visiting_right)

        recurse(first_child)
        if second_child:
            recurse(second_child)

    recurse(root_node_idx)


                 
    

            
def find_subsector_idx_for_position(level_data,
                                x, y,
                                node_callback=lambda node: None):
    nodes = level_data['NODES']
    node_idx = len(nodes)-1

    depth = 0
    while True:        
        if is_ssect_idx(node_idx):
            return get_real_idx(node_idx)
        else:
            node = nodes[get_real_idx(node_idx)]
            if is_on_left(node, x, y):
                node_callback(node, depth, on_left=True)
                node_idx = node.left_child
            else:
                node_callback(node, depth, on_left=False)
                node_idx = node.right_child
            depth += 1

            
    
def find_subsector_for_position(level_data,
                                x, y,
                                node_callback=lambda node: None):
    idx = find_subsector_idx_for_position(level_data, x ,y, node_callback)
    return level_data['SSECTORS'][idx]


def iterate_cur_pvs(level_data,
                    x, y,
                    cur_ssect_num,
                    node_callback = lambda node, visiting_left, visiting_right: None,
                    ssect_callback = lambda ssect_idx, ssect: None):
    
    global_pvs = level_data['PVS']

    cur_ssector = level_data['SSECTORS'][cur_ssect_num]
    
    cur_pvs = global_pvs[cur_ssect_num]

    def ssector_callback(ssect_idx, ssect):
        if ssect_idx in cur_pvs:
            ssect_callback(ssect_idx, ssect) 
        

    
    traverse_bsp_front_to_back(level_data,
                               x, y, cur_ssector,
                               node_callback,
                               ssect_callback=ssector_callback)

    
    #ssectors = level_data['SSECTORS'] 
    #for ssect_num in cur_pvs:
    #    ssect_callback(ssectors[ssect_num])
        
    
    

File: src/bsp.c
#include <genesis.h>

#include "draw_queues.h"
#include "level.h"


#include "math3d.h"

#include "span_buf.h"

// returns 1 if on-screen
int face_on_screen(Vect2D_s32 trans_v1, Vect2D_s32 trans_v2) {
    clip_result clip_res = clip_map_vertex(&trans_v1, &trans_v2);
    if(clip_res == OFFSCREEN) {
        return 0;
    }

    s16 screen_x1 = project_and_adjust_2d(trans_v1);
    s16 screen_x2 = project_and_adjust_2d(trans_v2);
    Line l; 
    l.pt1.x = screen_x1;
    l.pt1.y = 0; 
    l.pt2.x = screen_x2;
    l.pt2.y = 0;
    return BMP_clipLine(&l);
}

int bbox_on_screen(s16 left_x, s16 right_x, s16 top_y, s16 bottom_y) {
    Vect2D_s32 tl = transform_map_vert(left_x,  top_y);
    Vect2D_s32 tr = transform_map_vert(right_x, top_y);
    if(face_on_screen(tl, tr)) { return 1; }
    Vect2D_s32 bl = transform_map_vert(left_x,  bottom_y);
    if(face_on_screen(bl, tl)) { return 1; }
    Vect2D_s32 br = transform_map_vert(right_x, bottom_y);
    if(face_on_screen(bl, br)) { return 1; }
    return face_on_screen(tr, br);
}

void draw_face(Vect2D_s32 trans_v1, Vect2D_s32 trans_v2) {
    clip_result clip_res = clip_map_vertex(&trans_v1, &trans_v2);
    if(clip_res == OFFSCREEN) {
        return;
    }

    u8 col = (clip_res == UNCLIPPED) ? 0xFF : 0x11;

    transformed_vert screen_v1 = project_and_adjust_3d(trans_v1, -100, 100);
    transformed_vert screen_v2 = project_and_adjust_3d(trans_v2, -100, 100);
    
    Line l;
    l.col = col;
    l.pt1 = (Vect2D_s16){.x = screen_v1.x, .y = screen_v1.yceil};
    l.pt2 = (Vect2D_s16){.x = screen_v2.x, .y = screen_v2.yceil};
    if(BMP_clipLine(&l)) { BMP_drawLine(&l); }
    l.pt1 = (Vect2D_s16){.x = screen_v2.x, .y = screen_v2.yceil};
    l.pt2 = (Vect2D_s16){.x = screen_v2.x, .y = screen_v2.yfloor};
    if(BMP_clipLine(&l)) { BMP_drawLine(&l); }
    l.pt1 = (Vect2D_s16){.x = screen_v2.x, .y = screen_v2.yfloor};
    l.pt2 = (Vect2D_s16){.x = screen_v1.x, .y = screen_v1.yfloor};
    if(BMP_clipLine(&l)) { BMP_drawLine(&l); }
    l.pt1 = (Vect2D_s16){.x = screen_v1.x, .y = screen_v1.yfloor};
    l.pt2 = (Vect2D_s16){.x = screen_v1.x, .y = screen_v1.yceil};
    if(BMP_clipLine(&l)) { BMP_drawLine(&l); }

}

void draw_bbox(s16 left_x, s16 right_x, s16 top_y, s16 bottom_y) {
    
    // TODO: factor out shared calculations
    Vect2D_s32 tl = transform_map_vert(left_x,  top_y);
    Vect2D_s32 tr = transform_map_vert(right_x, top_y);
    Vect2D_s32 bl = transform_map_vert(left_x,  bottom_y);
    Vect2D_s32 br = transform_map_vert(right_x, bottom_y);

    draw_face(tl, tr);
    draw_face(tr, br);
    draw_face(br, bl);
    draw_face(bl, tl);

}

/*
void traverse_bsp_nodes_front_to_back(s32 x, s32 y) {
    node* root_node = &(cur_level->nodes[cur_level->num_nodes-1]);

    s16 left_x = root_node->left_box_left;
    s16 right_x = root_node->left_box_right;
    s16 top_y = root_node->left_box_top;
    s16 bottom_y = root_node->left_box_bottom;
    draw_bbox(left_x, right_x, top_y, bottom_y);
}
*/


int get_node_real_idx(u16 idx) {
    return (idx & (~(1<<15)));
}

int is_ssector_idx(u16 idx) {
    return idx & (1<<15);
}

int right_bbox_is_onscreen(const node* nd) {
    return bbox_on_screen(nd->right_box_left, nd->right_box_right, nd->right_box_top, nd->right_box_bottom);
}

int left_bbox_is_onscreen(const node* nd) {
    return bbox_on_screen(nd->left_box_left, nd->left_box_right, nd->left_box_top, nd->left_box_bottom);
}


int is_on_left(const node* nd, fix32 x, fix32 y) {
    s16 dx = fix32ToInt(x) - nd->split_x;
    s16 dy = fix32ToInt(y) - nd->split_y;

    return (((dx * nd->split_dy) - (dy * nd->split_dx)) <= 0);
}

ssector* find_subsector_for_position(fix32 x, fix32 y, int root_node_idx) {
    const node* nodes = cur_level->nodes;
    const ssector* subsectors = cur_level->ssectors;
      int recur(int node_idx) {
        if(is_ssector_idx(node_idx)) {
            return &subsectors[get_node_real_idx(node_idx)];
        } else {
            const node* nd = &(nodes[get_node_real_idx(node_idx)]);
            int on_left = is_on_left(nd, x, y);

            return recur(on_left ? nd->left_child : nd->right_child);
        }
    }
    
    recur(root_node_idx);
}


void draw_bsp_node(fix32 x, fix32 y, int root_node_idx) {
    const node* nodes = cur_level->nodes;

    int recur(int node_idx) {
        if(is_ssector_idx(node_idx)) {
            return enqueue_draw_ssector(get_node_real_idx(node_idx));
        } else {
            const node* nd = &(nodes[get_node_real_idx(node_idx)]);
            int on_left = is_on_left(nd, x, y);

            const uint16_t first_node_idx = (on_left ? nd->left_child : nd->right_child);
            const uint16_t second_node_idx = (on_left ? nd->right_child : nd->left_child);
            const u8 visiting_second_node = (on_left ? right_bbox_is_onscreen(nd) : left_bbox_is_onscreen(nd));

            u8 done = recur(first_node_idx);
            if(!done && visiting_second_node) {
                return recur(second_node_idx);
            }
            return done;
        }
    }
    
    recur(root_node_idx);
}